decomp2gef.py

```diff
@@ -362,37 +368,46 @@ def _construct_small_elf(self):
         self._elf_cache["data"] = open(self._elf_cache["fname"], "rb").read()
         return self._elf_cache["fname"]
 
-    def _force_update_text_size(self, elf_data, new_size):
-        # XXX: this is bad and only works on 64bit elf
-        default_text_size = 0x92
-        elf_rev = elf_data[::-1]
-        size_off = elf_rev.find(default_text_size)
-        patch = struct.pack("<Q", new_size)
-        size_off = len(elf_rev) - size_off - 1
+    def _force_update_text_size(self, stream, elf, elf_data, new_size):
+        text_sect = elf.get_section_by_name('.text')
+
+        for count in range(elf['e_shnum']):
+            sect_off = elf['e_shoff'] + count * elf['e_shentsize']
+            stream.seek(sect_off)
+            section_header = elf.structs.Elf_Shdr.parse_stream(stream)
+            if section_header['sh_name'] == text_sect['sh_name']:
+                break
 
-        elf_data[size_off:size_off + len(patch)] = patch
+        patch = struct.pack("<Q", new_size)
+        elf_data[sect_off+32: sect_off + 32 + len(patch)] = patch
         return elf_data
 
+
     def _force_update_sym_sizes(self, fname, queued_sym_sizes):
-        # parsing based on: https://github.com/torvalds/linux/blob/master/include/uapi/linux/elf.h
-        get_elf_headers.cache_clear()
-        elf: Elf = get_elf_headers(fname)
-        elf_data = bytearray(open(fname, "rb").read())
+        # use pyelftools to obtain accurate offsets and assign symbol sizes
+
+        stream = open(fname, 'rb')
+        elf_data = bytearray(stream.read())
+        stream.close()
+
+        stream = open(fname, 'rb')
+        elf = ELFFile(stream)
 
         # patch .text to seem large enough for any function
-        if elf.e_class == Elf.ELF_64_BITS:
-            elf_data = self._force_update_text_size(elf_data, 0xFFFFFF)
+        elf_data = self._force_update_text_size(stream, elf, elf_data, 0xFFFFFF)
 
         # find the symbol table
-        for section in elf.shdrs:
-            if section.sh_name == ".symtab":
-                break
-        else:
-            return
+        section = elf.get_section_by_name('.symtab')
+        if not section: return 
 
         # locate the location of the symbols size in the symtab
-        tab_offset = section.sh_offset
-        sym_data_size = 24 if elf.ELF_64_BITS else 16
+        tab_offset = section['sh_offset']
+        """
```

Comment:
comments in the middle of functions should all be "#" instead of the multiline comment style.